pkg/utils/tsoutil/tso_dispatcher_test.go

```diff
@@ -0,0 +1,195 @@
+			Logical:  1,                      // Logical timestamp is fixed at 1
+		},
+	}, nil
+}
+
+type mockProtoFactory struct {
+	stream *mockStream
+}
+
+func (m *mockProtoFactory) createForwardStream(_ context.Context, _ *grpc.ClientConn) (stream, context.CancelFunc, error) {
+	if m.stream == nil {
+		return nil, nil, errors.New("no stream available")
+	}
+	return m.stream, func() {}, nil
+}
+
+type mockRequest struct {
+	forwardedHost string
+	clientConn    *grpc.ClientConn
+	count         uint32
+	doneCh        chan struct{}
+	err           error
+}
+
+func (m *mockRequest) getForwardedHost() string {
+	return m.forwardedHost
+}
+
+func (m *mockRequest) getClientConn() *grpc.ClientConn {
+	return m.clientConn
+}
+
+func (m *mockRequest) getCount() uint32 {
+	return m.count
+}
+
+func (m *mockRequest) process(forwardStream stream, count uint32) (tsoResp, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	return forwardStream.process(0, count, 0, 0)
+}
+
+func (m *mockRequest) postProcess(countSum int64, _ int64, _ int64) (int64, error) {
+	close(m.doneCh)
+	return countSum, m.err
+}
+
+type tsoDispatcherTestSuite struct {
+	suite.Suite
+	dispatcher *TSODispatcher
+}
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m, testutil.LeakOptions...)
+}
+
+func TestTSODispatcherTestSuite(t *testing.T) {
+	suite.Run(t, new(tsoDispatcherTestSuite))
+}
+
+func (suite *tsoDispatcherTestSuite) SetupTest() {
+	tsoProxyHandleDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name: "test_tso_proxy_handle_duration_seconds",
+		Help: "Histogram of TSO proxy handle duration",
+	})
+	suite.dispatcher = NewTSODispatcher(tsoProxyHandleDuration, nil)
+}
+
+func (suite *tsoDispatcherTestSuite) TearDownTest() {
+	suite.dispatcher.Stop()
+	runtime.GC() // Force garbage collection to clean up any lingering goroutines
+}
+
+func (suite *tsoDispatcherTestSuite) TestGoroutineLeakOnStreamError() {
+	re := suite.Require()
+
+	// Create a mock tso stream
+	mockStream := &mockStream{
+		delay: 10 * time.Millisecond,
+	}
+	protoFactory := &mockProtoFactory{
+		stream: mockStream,
+	}
+
+	ctx := context.Background()
+	forwardedHost := "test-host"
+	var reqPendingCount atomic.Int64
+	var reqDispatchCount atomic.Int64
+	var wg sync.WaitGroup
+	// mock 11000 goroutines as tso proxy client
+	for range 11000 {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			dipatchReqDoneCh := make(chan struct{})
+			for range 10 {
+				req := &mockRequest{
+					forwardedHost: forwardedHost,
+					clientConn:    nil,
+					count:         1,
+					doneCh:        make(chan struct{}),
+				}
+
+				go func() {
+					// if there has goroutine leak, it will block here
+					suite.dispatcher.DispatchRequest(ctx, req, protoFactory)
+					dipatchReqDoneCh <- struct{}{}
+				}()
+
+				select {
+				case <-dipatchReqDoneCh:
+					reqDispatchCount.Add(1)
+				case <-time.After(20 * time.Second):
+					reqPendingCount.Add(1)
```

Comment:
It seems there is no need to introduce `reqPendingCount`; just check whether `reqDispatchCount` is 11000*10.